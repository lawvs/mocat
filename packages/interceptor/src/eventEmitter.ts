// See https://github.com/EventEmitter2/EventEmitter2
import { EventEmitter2 } from 'eventemitter2'
import type { EventAndListener } from 'eventemitter2'
import type { MockEventMap, RegisterEvent, RunEvent } from './types'

export interface MockEventEmitter {
  emit<T extends keyof MockEventMap>(event: T, value: MockEventMap[T]): boolean

  on<T extends keyof MockEventMap>(
    event: T,
    listener: (value: MockEventMap[T]) => void
  ): void
  once<T extends keyof MockEventMap>(
    event: T,
    listener: (value: MockEventMap[T]) => void
  ): void
  onAny<T extends keyof MockEventMap>(
    listener: (event: T, value: MockEventMap[T]) => void
  ): void
  off<T extends keyof MockEventMap>(
    event: T,
    listener: (...values: any[]) => void
  ): this
  offAny(listener: (value: any[]) => void): this
}

// @ts-ignore
export const eventEmitter: MockEventEmitter = new EventEmitter2()

export const debug = () => {
  const log: EventAndListener = (event, ...payload) =>
    // eslint-disable-next-line no-console
    console.log(
      `%cReceive%c %c${event.toString()}`,
      'background: rgba(0, 255, 255, 0.6); color: black; padding: 0px 6px; border-radius: 4px;',
      '',
      'text-decoration: underline',
      payload
    )
  eventEmitter.onAny(log)
  return () => eventEmitter.offAny(log)
}

// Register event

// const mockMap = new Map<any, any>()

export const registerMock = (rule: RegisterEvent) => {
  // mockMap.set(target, { desc, scenes })
  eventEmitter.emit(rule.type, rule)
}

// Run event

export const onRun = (event: RunEvent) => {
  eventEmitter.emit(event.type, event)
}
