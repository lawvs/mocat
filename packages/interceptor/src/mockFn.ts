import { registerMock, onRun } from './eventEmitter'
import type {
  AsyncFnAfterEvent,
  AsyncFnBeforeEvent,
  AsyncFnRegister,
} from './types'

export const mockAsyncFn = (
  mock: Omit<AsyncFnRegister, 'type' | 'target'> = {}
) => <T extends (...args: unknown[]) => Promise<unknown>>(targetFn: T): T => {
  const rule: AsyncFnRegister = {
    ...mock,
    type: 'Register/asyncFn',
    target: targetFn,
  }
  registerMock(rule)

  return ((...args: Parameters<T>) =>
    new Promise((resolve, reject) => {
      const beforeEvent: AsyncFnBeforeEvent = {
        type: 'Run/asyncFn/before',
        target: targetFn,
        rule,
        resolve,
        reject,
        pass: async (interceptReturn = false) => {
          let originResult: unknown
          let error: Error | undefined = undefined
          try {
            originResult = await targetFn(...args)
          } catch (e) {
            error = e
          }

          const passReturn = () => {
            if (error) {
              reject(error)
              return
            }
            resolve(originResult)
          }
          if (!interceptReturn) {
            passReturn()
            return
          }

          const afterEvent: AsyncFnAfterEvent = {
            type: 'Run/asyncFn/after',
            target: targetFn,
            rule,
            result: originResult,
            error,
            resolve,
            reject,
            pass: passReturn,
          }

          onRun(afterEvent)
        },
      }

      onRun(beforeEvent)
    })) as T
}
