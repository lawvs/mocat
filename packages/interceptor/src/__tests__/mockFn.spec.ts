import { mockAsync } from '../mockFn'
import { onAny } from '../eventEmitter'

describe('mockAsync should works as before mock', () => {
  onAny((_, payload) => payload.pass())

  test('mockAsync should works with resolve', async () => {
    const asyncFn = () => Promise.resolve(1)
    const mockFn = mockAsync()(asyncFn)
    expect(await mockFn()).toEqual(await asyncFn())
  })

  test('mockAsync should works with reject', async () => {
    const asyncFn = () => Promise.reject(1)
    const mockFn = mockAsync()(asyncFn)
    expect(await mockFn().catch((i) => i)).toEqual(
      await asyncFn().catch((i) => i)
    )
  })
})
