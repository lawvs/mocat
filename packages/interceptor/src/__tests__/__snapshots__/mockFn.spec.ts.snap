// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test event match snapshot 1`] = `
[
  {
    "desc": "it is a mock api",
    "name": "login api",
    "scenarios": [
      {
        "desc": "login success",
        "name": "scenario 1",
        "return": {
          "code": 0,
          "data": {
            "username": "admin",
          },
        },
      },
      {
        "desc": "login failure",
        "error": {
          "code": 1,
          "message": "password incorrect",
        },
        "name": "scenario 2",
      },
    ],
    "target": [Function],
    "timeStamp": 0,
    "type": "Register/asyncFn",
  },
]
`;

exports[`snapshot test event match snapshot 2`] = `
[
  {
    "pass": [Function],
    "reject": [Function],
    "resolve": [Function],
    "rule": {
      "desc": "it is a mock api",
      "name": "login api",
      "scenarios": [
        {
          "desc": "login success",
          "name": "scenario 1",
          "return": {
            "code": 0,
            "data": {
              "username": "admin",
            },
          },
        },
        {
          "desc": "login failure",
          "error": {
            "code": 1,
            "message": "password incorrect",
          },
          "name": "scenario 2",
        },
      ],
      "target": [Function],
      "timeStamp": 0,
      "type": "Register/asyncFn",
    },
    "target": [Function],
    "timeStamp": 0,
    "type": "Run/asyncFn/before",
  },
]
`;
